core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -316,7 +324,7 @@ class ReplicaManager(val config: KafkaConfig,
   private def maybeRemoveTopicMetrics(topic: String): Unit = {
     val topicHasOnlinePartition = allPartitions.values.exists {
       case HostedPartition.Online(partition) => topic == partition.topic
-      case HostedPartition.None | HostedPartition.Offline => false
+      case _ => false
```

Comment:
Using the wildcard means you won't get any help if you add any new HostedPartition values. Worth thinking if that's what we want as a general pattern or if we prefer to enumerate and/or extract to other methods.

Reply:
This is one case we probably _do_ want to act in the deferred state. One option I was consider is whether we should give 
```scala
sealed trait HostedPartition {
  def partition: Option[Partition]
}
```
Then we could refactor this as a `filter` or something.

Reply:
@hachikuji Have you considered having a sealed trait with `def partition: Partition` and then have the relevant case classes inherit from that? Then, when you want to get a partition, you match against that subtrait while knowing that the partition is always available in those cases. I suggested this offline to @rondagostino, but not sure if he didn't think it was useful or didn't get around to it.

Reply:
> a sealed trait with def partition: Partition

I like this solution better than making everything a case class with an `Option[Partition]`.  Will push a commit with it shortly.

Reply:
After this change, we can probably keep `HostedPartition.None | HostedPartition.Offline` as it was before, right?